Move AMP rendering of [youtube] shortcode to Jetpack from AMP plugin

kienstra opened the pull request:
#### Summary

This moves the AMP handling of the `[youtube]` shortcode from the [AMP plugin](https://github.com/ampproject/amp-wp/blob/b0bd40c18ce61741a366d54a9297dc440bf4330d/includes/embeds/class-amp-youtube-embed-handler.php#L138) to Jetpack. Very similar to #13921, but for `[youtube]` instead of `[vimeo]`.

Fixes https://github.com/ampproject/amp-wp/issues/3309

#### Changes proposed in this Pull Request:

* Apply the `[youtube]` shortcode handling [from the AMP plugin](https://github.com/ampproject/amp-wp/blob/b0bd40c18ce61741a366d54a9297dc440bf4330d/includes/embeds/class-amp-youtube-embed-handler.php#L138)
* Only refactor existing non-AMP `[youtube]` shortcode logic for reuse with the AMP shortcode, no intended change to it

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is a new feature for Jetpack, but moved from the AMP plugin

#### Testing instructions:
1. Ensure that `wp-config.php` has `define( 'JETPACK_DEV_DEBUG', true);`
2. In `/wp-admin`, in the Jetpack 'Settings' page, and in the 'Writing' tab, ensure this is toggled on:

<img width="818" alt="jetpack-here" src="https://user-images.githubusercontent.com/4063887/67992321-dded6400-fc01-11e9-81ea-9a05716d1eca.png">

3. Fetch this PR's branch
4. Clone the AMP plugin 
```
$ git clone --recursive https://github.com/ampproject/amp-wp.git amp && cd amp
```
5. Fetch this PR of the AMP plugin, which deactivates the AMP plugin's `[youtube]` shortcode callback: https://github.com/ampproject/amp-wp/pull/3678
6. Do `$ npm install && composer install`
7. Activate the AMP plugin
8. Create a new post with a Shortcode block
9. Add a shortcode, like:
```
[youtube https://www.youtube.com/watch?v=_Oh12ROTQCE]
```
10. Preview the front-end of the AMP URL, like by adding `&amp` or `?amp`to the URL, and look at how the Vimeo shortcode looks
11. `checkout` the `master` branch of Jetpack, not this PR's branch
12. Preview the front-end of the AMP URL, and notice how the Vimeo shortcode looks
13. Expected: The shortcode should look the same, whether using this PR's branch, or the Jetpack plugin's `master` branch
14. Test more shortcodes under the [More Examples heading here](https://en.support.wordpress.com/videos/youtube/)

## Before

<img width="896" alt="amp-pr-before" src="https://user-images.githubusercontent.com/4063887/68559299-1a8b3d80-0402-11ea-90a8-ca536f97f668.png">

## After

<img width="859" alt="amp-pr-after" src="https://user-images.githubusercontent.com/4063887/68559306-20811e80-0402-11ea-8d62-270721105028.png">

#### Proposed changelog entry for your changes:
Migrate AMP-conversion of `[youtube]` shortcode to Jetpack from the AMP plugin

kienstra committed:
Move the AMP rendering of [youtube] shortcode to Jetpack

In continuation of the [vimeo] shortcode
AMP migration.
The intent is to not change how the rendering
worked in the AMP plugin,
only to migrate it to Jetpack.

kienstra committed:
Add a trailing , to end of array()

To follow coding standards,
ensure the array() ends with a ,

kienstra committed:
Correct failed unit tests for 'no_attributes'

This should actually be '',
as that is the behavior of the AMP plugin before.

kienstra commented:
**Separate PRs?**

Hi @jeherve,
Hope you had a great weekend.

Is it alright to open a separate PR for each shortcodes to migrate to Jetpack?

If it's alright, there are [4 more AMP shortcodes](https://github.com/ampproject/amp-wp/issues/3309#issue-495949833) after this that we'd like to migrate to Jetpack. 

Though this isn't urgent, I know reviewing them can be a lot of work.

Also, let me know if you'd like me to fix the [Code Climate issues](https://codeclimate.com/github/Automattic/jetpack/pull/14010).

Thanks, Jeremy!

jeherve commented:
> Is it alright to open a separate PR for each shortcodes to migrate to Jetpack?

That's absolutely fine, yes. In fact, separate PRs are better I think, they're easier to review and to work on.

jeherve commented on modules/shortcodes/youtube.php:
```diff
@@ -365,16 +294,203 @@ function youtube_id( $url ) {
+ *
+ * @return string|bool The video ID based on that URL, or false.
+ */
+function get_amp_youtube_id_from_url( $url ) {
```
We currently have a `jetpack_get_youtube_id` function one can use to get an ID from a YouTube URL. Do you think it would be helpful to you here, or are there changes from `get_amp_youtube_id_from_url` you think we should bring to `jetpack_get_youtube_id`? In any case, I think it'd be nice if we only had one function to take care of that.

https://github.com/Automattic/jetpack/blob/792b26b5539d07cc35fdd93567942f2ad481eef2/functions.compat.php#L20

kienstra replied:
Great point, I didn't see that function.

kienstra replied:
https://github.com/Automattic/jetpack/pull/14010/commits/d7c05647435878c32ee1bece9814c4c933bad32e uses `jetpack_get_youtube_id()` instead of the custom function ported from AMP. 

There doesn't look to be an regression from this. It even passed this unit test when substituting `jetpack_get_youtube_id` for `get_amp_youtube_id_from_url()` on [this line](https://github.com/Automattic/jetpack/pull/14010/commits/d7c05647435878c32ee1bece9814c4c933bad32e#diff-7835daad42d569641fcb9ae94cce9f51L280).

kienstra committed:
Remove custom YouTube ID function in favor of existing one

As Jeremy mentioned,
jetpack_get_youtube_id() should work
in place of the custom function
get_amp_youtube_id_from_url().
It actually passes the unit test:
test_get_amp_youtube_id_from_url().

jeherve commented on modules/shortcodes/youtube.php:
```diff
@@ -365,16 +294,162 @@ function youtube_id( $url ) {
+	) {
+		return jetpack_amp_youtube_shortcode( $atts );
+	} else {
+		return youtube_id( ( isset( $atts[0] ) ) ? ltrim( $atts[0], '=' ) : shortcode_new_to_old_params( $atts ) );
```
What do you think about setting the URL at the beginning of the `youtube_shortcode` function, so you don't have to repeat that `isset( $atts[0] ) ) ? ltrim( $atts[0], '=' ) : shortcode_new_to_old_params( $atts )` twice (here and in the `jetpack_amp_youtube_shortcode` function)?

kienstra replied:
That's a good idea. Is https://github.com/Automattic/jetpack/pull/14010/commits/51a5eb78c534d1c9b7c2edef2718b04d167fb5f4 what you had in mind?

kienstra replied:
That's a good idea, it should be more DRY. How does https://github.com/Automattic/jetpack/pull/14010/commits/51a5eb78c534d1c9b7c2edef2718b04d167fb5f4 look?

kienstra committed:
Close the fallback link

As Jeremy mentioned,
the <a> with the fallback URL
didn't have a closing </a> before.

kienstra commented on modules/shortcodes/youtube.php:
```diff
@@ -365,16 +294,162 @@ function youtube_id( $url ) {
+	$video_id = jetpack_get_youtube_id( $url );
+	if ( empty( $video_id ) ) {
+		return sprintf(
+			'<a href="%s" class="amp-wp-embed-fallback">',
```
Good point, https://github.com/Automattic/jetpack/pull/14010/commits/08c760b66e61ef4bbc1ce1dda96ff53f3e2c8d7f closes it.

kienstra committed:
Only run URL logic once

As Jeremy mentioned,
this logic was run in 2 functions.
So move this to the top of youtube_shortcode().

kienstra committed:
Rename jetpack_shortcode_youtube_query_args()

As Jeremy mentioned,
it doesn't only get query args.

kienstra commented on modules/shortcodes/youtube.php:
```diff
@@ -181,87 +181,16 @@ function youtube_id( $url ) {
-	if ( ! empty( $url['query'] ) ) {
-		wp_parse_str( $url['query'], $qargs );
-	} else {
+	$qargs = jetpack_shortcode_youtube_query_args( $url );
```
Yeah, that's a good idea. It's renamed in https://github.com/Automattic/jetpack/pull/14010/commits/eb7c5e77afe069463ded3731764e65aa51f0d839.

kienstra committed:
Return an error comment if the $args are empty, like above

As Jeremy mentioned,
if there's no $id, there's a comment
with an error message.
So do something similar if the $args are empty.

kienstra commented on modules/shortcodes/youtube.php:
```diff
@@ -181,87 +181,16 @@ function youtube_id( $url ) {
-	} else {
+	$qargs = jetpack_shortcode_youtube_query_args( $url );
+	if ( empty( $qargs ) ) {
 		return false;
```
Good idea. How does https://github.com/Automattic/jetpack/pull/14010/commits/bed7e2cd4ce363849e6b35b419342db965082b4a look?

jeherve commented on modules/shortcodes/youtube.php:
```diff
@@ -365,16 +294,159 @@ function youtube_id( $url ) {
+	$video_id = jetpack_get_youtube_id( $url );
+	if ( empty( $video_id ) ) {
+		return sprintf(
+			'<a href="%s" class="amp-wp-embed-fallback"></a>',
```
What do you think about adding something to the link to be more descriptive? Even the URL alone may be nice I think.

```suggestion
			'<a href="%s" class="amp-wp-embed-fallback">%s</a>',
```

kienstra replied:
That's a good idea.

kienstra replied:
Ah, I might need to update a test for that.

jeherve reviewed (changes requested):
Nice edits, we are getting there I think. I only have one remark right now.

kienstra committed:
Commit Jeremy's suggestion to add the URL as a child of the <a>

Co-Authored-By: Jeremy Herve <jeremy@tagada.hu>

kienstra committed:
Update unit tests for the previous commit

Also, pass another argument to sprintf().
And update a few tests.

jeherve reviewed (dismissed):
This looks good to me. I'll merge on Monday after testing on WordPress.com as well. Thank you!

kienstra commented:
Thanks, Jeremy! Have a great weekend.

jeherve committed:
Skip AMP tests on WordPress.com

WordPress.com does not run the latest version of the AMP plugin yet.
